File: server/internal/platform/setup/sampledata/mfg.go
package sampledata

import (
	"context"

	"github.com/google/uuid"
	"github.com/z1coyan/synie/server/internal/domain/manufacturing/master"
	"github.com/z1coyan/synie/server/internal/platform/authz"
)

func seedMfg(ctx context.Context, deps Dependencies, actor *authz.Actor, md masterData) (mfgResult, error) {
	opsByName := map[string]uuid.UUID{}
	var opIDs []uuid.UUID
	for _, name := range []string{"下料", "冲压", "折弯", "喷涂", "装配"} {
		op, err := deps.ManufacturingMaster.CreateOperation(ctx, actor, master.HeadCreateInput{Name: name})
		if err != nil {
			return mfgResult{}, err
		}
		opsByName[name] = op.ID
		opIDs = append(opIDs, op.ID)
	}

	t1, err := deps.ManufacturingMaster.CreateTemplate(ctx, actor, master.HeadCreateInput{Name: "钣金件标准工艺"})
	if err != nil {
		return mfgResult{}, err
	}
	for _, row := range []struct {
		op           string
		seq          int64
		req          string
		isOutsourced bool
	}{
		{"下料", 10, "按图下料,去毛刺", false},
		{"冲压", 20, "冲孔/落料一次成型", false},
		{"折弯", 30, "按图折弯,角度±1°", false},
		{"喷涂", 40, "外协喷涂,RAL7035", true},
	} {
		if _, err := deps.ManufacturingMaster.CreateTemplateItem(ctx, actor, t1.ID, master.RouteItemInput{
			Seq: row.seq, Requirement: ptr(row.req), IsOutsourced: row.isOutsourced,
			OperationID: opsByName[row.op],
		}); err != nil {
			return mfgResult{}, err
		}
	}

	t2, err := deps.ManufacturingMaster.CreateTemplate(ctx, actor, master.HeadCreateInput{Name: "铜排组件工艺"})
	if err != nil {
		return mfgResult{}, err
	}
	for _, row := range []struct {
		op  string
		seq int64
		req string
	}{
		{"下料", 10, "铜排定尺下料"},
		{"冲压", 20, "冲安装孔,去毛刺"},
		{"装配", 30, "端子压接,扭力按规范"},
	} {
		if _, err := deps.ManufacturingMaster.CreateTemplateItem(ctx, actor, t2.ID, master.RouteItemInput{
			Seq: row.seq, Requirement: ptr(row.req), OperationID: opsByName[row.op],
		}); err != nil {
			return mfgResult{}, err
		}
	}

	// BOM1 配电箱壳体 + 钣金模板路线
	bom1, err := deps.ManufacturingMaster.CreateBOM(ctx, actor, master.BOMCreateInput{
		MaterialID: md.Materials["box_shell"].ID, Note: ptr("示例 BOM"),
	})
	if err != nil {
		return mfgResult{}, err
	}
	for _, row := range []struct {
		key  string
		qty  string
		loss *string
		note *string
	}{
		{"steel_sheet", "2.5", nil, ptr("箱体展开料")},
		{"screw", "12", ptr("0.02"), ptr("装配紧固")},
		{"insul_sleeve", "0.5", nil, nil},
	} {
		if err := createBOMComponent(ctx, deps, actor, bom1.ID, md, row.key, row.qty, row.loss, row.note); err != nil {
			return mfgResult{}, err
		}
	}
	if _, err := deps.ManufacturingMaster.ApplyRouteTemplate(ctx, actor, bom1.ID, t1.ID); err != nil {
		return mfgResult{}, err
	}

	// BOM2 汇流铜排组件 + 手工路线 + 副产品
	bom2, err := deps.ManufacturingMaster.CreateBOM(ctx, actor, master.BOMCreateInput{
		MaterialID: md.Materials["busbar"].ID, Note: ptr("示例 BOM"),
	})
	if err != nil {
		return mfgResult{}, err
	}
	for _, row := range []struct {
		key  string
		qty  string
		loss *string
	}{
		{"copper_bar", "1.2", ptr("0.03")},
		{"terminal_block", "8", nil},
		{"insul_sleeve", "0.3", nil},
	} {
		if err := createBOMComponent(ctx, deps, actor, bom2.ID, md, row.key, row.qty, row.loss, nil); err != nil {
			return mfgResult{}, err
		}
	}
	for _, row := range []struct {
		op  string
		seq int64
		req string
	}{
		{"下料", 10, "铜排定尺下料"},
		{"装配", 20, "端子压接"},
	} {
		if _, err := deps.ManufacturingMaster.CreateBOMRoute(ctx, actor, bom2.ID, master.RouteItemInput{
			Seq: row.seq, Requirement: ptr(row.req), OperationID: opsByName[row.op],
		}); err != nil {
			return mfgResult{}, err
		}
	}
	scrap := md.Materials["scrap_copper"]
	if _, err := deps.ManufacturingMaster.CreateBOMByproduct(ctx, actor, master.ByproductInput{
		Quantity: dec("0.05"), Note: ptr("下料边角料"), BOMID: bom2.ID,
		MaterialID: scrap.ID, UnitID: scrap.DefaultUnitID,
	}); err != nil {
		return mfgResult{}, err
	}

	return mfgResult{
		Operations: opIDs, ProcessTemplates: []uuid.UUID{t1.ID, t2.ID},
		BOMs: []uuid.UUID{bom1.ID, bom2.ID}, OpsByName: opsByName,
	}, nil
}

func createBOMComponent(
	ctx context.Context, deps Dependencies, actor *authz.Actor, bomID uuid.UUID, md masterData,
	key, qty string, loss, note *string,
) error {
	mat := md.Materials[key]
	input := master.ComponentInput{
		Quantity: dec(qty), Note: note, BOMID: bomID,
		MaterialID: mat.ID, UnitID: mat.DefaultUnitID,
	}
	if loss != nil {
		v := dec(*loss)
		input.LossRate = &v
	}
	_, err := deps.ManufacturingMaster.CreateBOMComponent(ctx, actor, input)
	return err
}

File: server/src/modules/setup/sampledata/mfg.ts
import type { Actor } from '~/platform/authz/actor.ts'
import type { MasterData } from './helpers.ts'
import type { MfgResult, SampleDataDeps } from './types.ts'

export async function seedMfg(
  deps: SampleDataDeps,
  actor: Actor,
  md: MasterData,
): Promise<MfgResult> {
  const opsByName: Record<string, string> = {}
  const opIDs: string[] = []
  for (const name of ['下料', '冲压', '折弯', '喷涂', '装配']) {
    const op = await deps.manufacturingMaster.createOperation(actor, { name })
    opsByName[name] = op.id
    opIDs.push(op.id)
  }

  const t1 = await deps.manufacturingMaster.createTemplate(actor, { name: '钣金件标准工艺' })
  for (const row of [
    { op: '下料', seq: 10, req: '按图下料,去毛刺', isOutsourced: false },
    { op: '冲压', seq: 20, req: '冲孔/落料一次成型', isOutsourced: false },
    { op: '折弯', seq: 30, req: '按图折弯,角度±1°', isOutsourced: false },
    { op: '喷涂', seq: 40, req: '外协喷涂,RAL7035', isOutsourced: true },
  ] as const) {
    await deps.manufacturingMaster.createTemplateItem(actor, {
      templateId: t1.id,
      operationId: opsByName[row.op]!,
      seq: row.seq,
      requirement: row.req,
      isOutsourced: row.isOutsourced,
    })
  }

  const t2 = await deps.manufacturingMaster.createTemplate(actor, { name: '铜排组件工艺' })
  for (const row of [
    { op: '下料', seq: 10, req: '铜排定尺下料' },
    { op: '冲压', seq: 20, req: '冲安装孔,去毛刺' },
    { op: '装配', seq: 30, req: '端子压接,扭力按规范' },
  ] as const) {
    await deps.manufacturingMaster.createTemplateItem(actor, {
      templateId: t2.id,
      operationId: opsByName[row.op]!,
      seq: row.seq,
      requirement: row.req,
    })
  }

  const bom1 = await deps.manufacturingMaster.createBom(actor, {
    materialId: md.materials.box_shell!.id,
    note: '示例 BOM',
  })
  for (const row of [
    { key: 'steel_sheet', qty: '2.5', loss: null as string | null, note: '箱体展开料' },
    { key: 'screw', qty: '12', loss: '0.02', note: '装配紧固' },
    { key: 'insul_sleeve', qty: '0.5', loss: null, note: null },
  ]) {
    await createBOMComponent(deps, actor, bom1.id, md, row.key, row.qty, row.loss, row.note)
  }
  await deps.manufacturingMaster.applyRouteTemplate(actor, bom1.id, t1.id)

  const bom2 = await deps.manufacturingMaster.createBom(actor, {
    materialId: md.materials.busbar!.id,
    note: '示例 BOM',
  })
  for (const row of [
    { key: 'copper_bar', qty: '1.2', loss: '0.03' as string | null },
    { key: 'terminal_block', qty: '8', loss: null },
    { key: 'insul_sleeve', qty: '0.3', loss: null },
  ]) {
    await createBOMComponent(deps, actor, bom2.id, md, row.key, row.qty, row.loss, null)
  }
  for (const row of [
    { op: '下料', seq: 10, req: '铜排定尺下料' },
    { op: '装配', seq: 20, req: '端子压接' },
  ] as const) {
    await deps.manufacturingMaster.createRoute(actor, {
      bomId: bom2.id,
      operationId: opsByName[row.op]!,
      seq: row.seq,
      requirement: row.req,
    })
  }
  const scrap = md.materials.scrap_copper!
  await deps.manufacturingMaster.createByproduct(actor, {
    bomId: bom2.id,
    quantity: '0.05',
    note: '下料边角料',
    materialId: scrap.id,
    unitId: scrap.defaultUnitId,
  })

  return {
    operations: opIDs,
    processTemplates: [t1.id, t2.id],
    boms: [bom1.id, bom2.id],
    opsByName,
  }
}

export async function createBOMComponent(
  deps: SampleDataDeps,
  actor: Actor,
  bomId: string,
  md: MasterData,
  key: string,
  qty: string,
  loss: string | null,
  note: string | null,
): Promise<void> {
  const mat = md.materials[key]
  if (!mat) throw new Error(`示例物料缺失: ${key}`)
  await deps.manufacturingMaster.createComponent(actor, {
    bomId,
    materialId: mat.id,
    unitId: mat.defaultUnitId,
    quantity: qty,
    lossRate: loss,
    note,
  })
}
